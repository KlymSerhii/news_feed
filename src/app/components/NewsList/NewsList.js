import React, {useState, useMemo} from "react";
import {useDispatch, useSelector} from "react-redux";
import {makeStyles} from "@material-ui/core/styles";
import Button from "@material-ui/core/Button";
import NewsItem from "../NewsItem"
import Dialog from "../Dialog";
import {selectNews, deleteNews} from "../../slices/newsSlice";
import {selectUserStatus} from "../../slices/userSlice";

const generateDeleteNewsDialogData = (id) => ({
    title: "You are going to delete important news",
    text: `News id is ${id}. Are you sure?`,
    agreeText: "Yes",
    disagreeText: "No"
});

const useStyles = makeStyles((theme) => ({
    root: {
        display: "flex",
        flexGrow: 1,
        flexDirection: "row",
        justifyContent: "flex-start",
        flexFlow: "column wrap",
        marginTop: theme.spacing(2)
    },
    backdrop: {
        zIndex: theme.zIndex.drawer + 1,
        color: "#fff",
    },
}));

const NewsList = () => {
    const classes = useStyles();
    const [newsToDeleteData, setNewsToDeleteData] = useState(null);
    const news = useSelector(selectNews);
    const isUserLoggedIn = useSelector(selectUserStatus);
    const dispatch = useDispatch();
    const onDeleteNewsItem = (id) => () => setNewsToDeleteData({id, ...generateDeleteNewsDialogData(id)});
    const onAgreeDeleteNewsItem = () => {
        dispatch(deleteNews(newsToDeleteData.id));
        setNewsToDeleteData(null)
    };
    const onDisagreeDeleteNewsItem = () => setNewsToDeleteData(null);

    const newsItems = news.map(({id, title, text}) => (
        <NewsItem key={id} title={title} text={text} onDelete={onDeleteNewsItem(id)} isUserLoggedIn={isUserLoggedIn}/>));
    const dialogActions = useMemo(() => (<>
        <Button onClick={onDisagreeDeleteNewsItem} color="primary">
            Disagree
        </Button>
        <Button onClick={onAgreeDeleteNewsItem} color="primary" autoFocus>
            Agree
        </Button>
    </>), [onAgreeDeleteNewsItem, onDisagreeDeleteNewsItem]);

    return (
        <div className={classes.root}>
            {newsItems}
            {
                newsToDeleteData !== null &&
                <Dialog text={newsToDeleteData.text} title={newsToDeleteData.title} actions={dialogActions}/>
            }
        </div>
    );
};

export default NewsList
