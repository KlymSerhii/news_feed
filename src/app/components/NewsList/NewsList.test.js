import React from "react";
import { render, act, fireEvent } from "@testing-library/react"
import {useSelector} from "react-redux";
import NewsList from "./NewsList";

jest.mock("react-redux", () => ({
    useDispatch: jest.fn(),
    useSelector: jest.fn(),
}));

useSelector.mockImplementation((selector) => selector({
        user: {
            isLoggedIn: true
        },
        news: {
            items: [
                {
                    "id": 1,
                    "title": "First news",
                    "text": "First news goes first"
                },
                {
                    "id": 2,
                    "title": "Second news",
                    "text": "Second news goes second"
                },
                {
                    "id": 3,
                    "title": "Third news",
                    "text": "Third news goes third"
                },
                {
                    "id": 4,
                    "title": "Fourth news",
                    "text": "Fourth news goes fourth"
                },
                {
                    "id": 5,
                    "title": "Fifth news",
                    "text": "Fifth news goes fifth"
                },
                {
                    "id": 6,
                    "title": "Sixth news",
                    "text": "Sixth news goes sixth"
                }
            ]
        }
    }
));

describe("AddNews", () => {

    it("matches snapshot", () => {
        const { container } = render(<NewsList/>);
        expect(container).toMatchSnapshot();
    });

    it("renders 6 news items", () => {
        const { container } = render(<NewsList/>);
        expect(container.querySelectorAll(".MuiPaper-root").length).toBe(6);
    });

    it("Shows delete modal after user clicks on delete button", async () => {
        const { container } = render(<NewsList/>);
        const secondDeleteButton = container.querySelectorAll("button")[2];
        await act(async () => {
            fireEvent(
                secondDeleteButton,
                new MouseEvent("click", {
                    bubbles: true,
                    cancelable: true,
                })
            )
        });
        expect(container.querySelectorAll("#alert-dialog-title")).toBeTruthy();
    });
});
