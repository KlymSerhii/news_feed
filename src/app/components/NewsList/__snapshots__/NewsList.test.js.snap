// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddNews matches snapshot 1`] = `
<div>
  <div
    class="makeStyles-root-1"
  >
    <div
      class="MuiPaper-root MuiCard-root makeStyles-root-3 MuiPaper-outlined MuiPaper-rounded"
    >
      <div
        class="MuiCardContent-root"
      >
        <h2
          class="MuiTypography-root MuiTypography-h5"
        >
          First news
        </h2>
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          First news goes first
        </p>
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-deleteButton-4 MuiButton-textSecondary MuiButton-textSizeSmall MuiButton-sizeSmall"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Delete
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="MuiPaper-root MuiCard-root makeStyles-root-3 MuiPaper-outlined MuiPaper-rounded"
    >
      <div
        class="MuiCardContent-root"
      >
        <h2
          class="MuiTypography-root MuiTypography-h5"
        >
          Second news
        </h2>
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          Second news goes second
        </p>
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-deleteButton-4 MuiButton-textSecondary MuiButton-textSizeSmall MuiButton-sizeSmall"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Delete
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="MuiPaper-root MuiCard-root makeStyles-root-3 MuiPaper-outlined MuiPaper-rounded"
    >
      <div
        class="MuiCardContent-root"
      >
        <h2
          class="MuiTypography-root MuiTypography-h5"
        >
          Third news
        </h2>
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          Third news goes third
        </p>
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-deleteButton-4 MuiButton-textSecondary MuiButton-textSizeSmall MuiButton-sizeSmall"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Delete
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="MuiPaper-root MuiCard-root makeStyles-root-3 MuiPaper-outlined MuiPaper-rounded"
    >
      <div
        class="MuiCardContent-root"
      >
        <h2
          class="MuiTypography-root MuiTypography-h5"
        >
          Fourth news
        </h2>
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          Fourth news goes fourth
        </p>
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-deleteButton-4 MuiButton-textSecondary MuiButton-textSizeSmall MuiButton-sizeSmall"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Delete
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="MuiPaper-root MuiCard-root makeStyles-root-3 MuiPaper-outlined MuiPaper-rounded"
    >
      <div
        class="MuiCardContent-root"
      >
        <h2
          class="MuiTypography-root MuiTypography-h5"
        >
          Fifth news
        </h2>
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          Fifth news goes fifth
        </p>
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-deleteButton-4 MuiButton-textSecondary MuiButton-textSizeSmall MuiButton-sizeSmall"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Delete
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="MuiPaper-root MuiCard-root makeStyles-root-3 MuiPaper-outlined MuiPaper-rounded"
    >
      <div
        class="MuiCardContent-root"
      >
        <h2
          class="MuiTypography-root MuiTypography-h5"
        >
          Sixth news
        </h2>
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          Sixth news goes sixth
        </p>
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-deleteButton-4 MuiButton-textSecondary MuiButton-textSizeSmall MuiButton-sizeSmall"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Delete
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
